apps/emqx_ft/src/emqx_ft_storage_exporter_fs.erl

```diff
@@ -0,0 +1,333 @@
+    meta :: filemeta(),
+    hash :: crypto:hash_state()
+}).
+
+-opaque export() :: #export{}.
+
+-type reader() :: pid().
+
+-define(TEMPDIR, "tmp").
+-define(MANIFEST, ".MANIFEST.json").
+
+%% NOTE
+%% Bucketing of resulting files to accomodate the storage backend for considerably
+%% large (e.g. > 10s of millions) amount of files.
+-define(BUCKET_HASH, sha).
+
+%% 2 symbols = at most 256 directories on the upper level
+-define(BUCKET1_LEN, 2).
+%% 2 symbols = at most 256 directories on the second level
+-define(BUCKET2_LEN, 2).
+
+%%
+
+-spec start_export(options(), transfer(), filemeta()) ->
+    {ok, export()} | {error, file_error()}.
+start_export(Options, Transfer, Filemeta = #{name := Filename}) ->
+    TempFilepath = mk_temp_filepath(Options, Transfer, Filename),
+    ResultFilepath = mk_filepath(Options, Transfer, result, Filename),
+    _ = filelib:ensure_dir(TempFilepath),
+    case file:open(TempFilepath, [write, raw, binary]) of
+        {ok, Handle} ->
+            {ok, #export{
+                path = TempFilepath,
+                handle = Handle,
+                result = ResultFilepath,
+                meta = Filemeta,
+                hash = init_checksum(Filemeta)
+            }};
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec write(export(), iodata()) ->
+    {ok, export()} | {error, file_error()}.
+write(Export = #export{handle = Handle, hash = Ctx}, IoData) ->
+    case file:write(Handle, IoData) of
+        ok ->
+            {ok, Export#export{hash = update_checksum(Ctx, IoData)}};
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec complete(export()) ->
+    ok | {error, {checksum, _Algo, _Computed}} | {error, file_error()}.
+complete(
+    Export = #export{
+        path = Filepath,
+        handle = Handle,
+        result = ResultFilepath,
+        meta = FilemetaIn,
+        hash = Ctx
+    }
+) ->
+    case verify_checksum(Ctx, FilemetaIn) of
+        {ok, Filemeta} ->
+            ok = file:close(Handle),
+            _ = filelib:ensure_dir(ResultFilepath),
+            _ = file:write_file(mk_manifest_filepath(ResultFilepath), encode_filemeta(Filemeta)),
+            file:rename(Filepath, ResultFilepath);
+        {error, _} = Error ->
+            _ = discard(Export),
+            Error
+    end.
+
+-spec discard(export()) ->
+    ok.
+discard(#export{path = Filepath, handle = Handle}) ->
+    ok = file:close(Handle),
+    file:delete(Filepath).
+
+%%
+
+-spec list_local(options(), transfer()) ->
+    {ok, [exportinfo(), ...]} | {error, file_error()}.
+list_local(Options, Transfer) ->
+    Dirpath = mk_filedir(Options, Transfer, result),
+    case
+        emqx_ft_fs_util:fold(
+            fun
+                (_Path, {error, Reason}, [], []) ->
+                    {error, Reason};
+                (_Path, Fileinfo = #file_info{type = regular}, [Filename | _], Acc) ->
+                    Filepath = filename:join(mk_result_dirname(Transfer) ++ [Filename]),
```

Comment:
Yeah, I suspected it will be confusing. `mk_filepath` gives absolute filepath, while here we need filepath _relative to the exporter storage root_. Maybe clearer naming / more descriptive type aliases will help?